src/common/options/rgw.yaml.in

```diff
@@ -3850,3 +3850,21 @@ options:
   services:
   - rgw
   with_legacy: true
+- name: rgw_persistent_topic_max_queue_size
+  type: size
+  level: advanced
+  desc: The maximum size (in bytes) of rados queue that stores the persistent notifications.
+  default: 128_M
+  services:
+  - rgw
+  flags:
+  - startup
+  with_legacy: true
+- name: rgw_notification_max_workers
```

Comment:
we should probably first check if changing the value increase the notification throughput.
* the major cost in enabling notifications is when writing the notifications one-by-one to the queue (notification rados operation for each S3 operation)
* reading from the queue and sending the messages is done in large bulks, so the ratio is much better
* having multiple workers reading from queues without queue affinity will probably result in contention and would not improve performance

Reply:
Regarding measuring the throughput, we are actually keen to know the notification latencies.. is there any metric that we can look at?
other option we were think was to log every successful message posted to broker and print the latency along with it (just like beast log which logs the latency for every successful s3 request)
and then we can pull those logs and plot the graph for average_latencies on it ?

Reply:
On comment regarding the `the major cost in enabling notifications is when writing the notifications one-by-one to the queue`,
you mean increasing this value could cause contention ?
i thought this value is only used for while we are processing queue, writing to queue is still handled by same rgw thread that handles that s3 operation ?

In our large scale testing, dispatching notification to broker could not keep up with the rate at which s3 messages were written (notification were written to queue), eventually causing the queue to get full
Also in current implementation of dispatching, currently single rgw instances once locks the queue will do the dispatching !

Reply:
> On comment regarding the comment `the major cost in enabling notifications is when writing the notifications one-by-one to the queue`, you mean increasing the thread values could cause contention ? i thought this value is only used for processing queue, writing to queue is still handled by same rgw thread that handles that s3 operation ?
> 

yes, you are right. this value does not impact writing to the queue, only reading.
it may cause contention on the readers. since only one reader can read from the queue at a time, and the readers does not do much other than reading - they read from the queue, figure out to which subsystem the events should go to (kafka or amqp) and put them in an in-memory, lockfree queue to these systems,
they do the reading from the RADOS queue in large bulks, so this is quite efficient, and unlikely that the bottleneck will be there.

> In our large scale testing, dispatching notification to broker could not keep up with the rate at which s3 messages were written (notification were written to queue), eventually causing the queue to get full 

dispatching the messages to the broker is at the responsibility of the amqp/kafka worker threads (not the notifications). note that, unlike S3, the amqp/kafka protocols are async binary protocols. again. unlikely that this is the bottleneck, but if it is, we would probably need to add more worker threads there.
note that if the broker cannot keep up with the message rate, giving the amqp/kafka client more threads will not help.

> Also in current implementation of dispatching, currently single rgw instances once locks the queue will do the dispatching !

the overall assumptions is that we will not have one topic to serve all notification from all buckets, and different topic queues will be owned by different RGWs.
but even in a very unbalanced case, the RGW that writes the notification is doing one RADOS op per S3 op, while the RGW that reads the notifications and dispatch them is doing one RADOS op per 1000 S3 ops.

Reply:
> @yuvalif Regarding measuring the throughput, we are actually keen to know the notification latencies.. is there any metric that we can look at? other option we were think was to log every successful message posted to broker and print the latency along with it (just like beast log which logs the latency for every successful s3 request) and then we can pull those logs and plot the graph for average_latencies on it ?

the notification latency will be highly impacted by the latency of the broker, so i'm not sure how useful that number would be for tuning the RGW.
adding logs to the RGW will allow you to measure that, but would have its own negative impact on the performance and latency.
simple option would be to measure that on the subscriber side. the subscriber could compare the event timestamp to the current time when it reads the notification. however, this would include the time it takes the broker to send the notifications to the subscriber.
to measure this more accuratly on the RGW, we should add that to the topic stats. if we decide this is a priority, I would recomment adding 2 (or 4) numbers:
* median (and 95th percentile) latency inside ceph
* median (and 95th percentile) end2end latency
as with any gauge values we would need to decide when they are reset.

note that, overall, latency (within reasonable limits) was never a goal for bucket notifications, and there are no guarantees on that (even in AWS). application developers should not use bucket notifications for latency sensitive applications.